client/modules/IDE/components/ConsoleInput.jsx

```diff
@@ -47,92 +41,84 @@ class ConsoleInput extends React.Component {
-        const lineNumber = this._cm.getDoc().getCursor().line;
+        const lineNumber = cm.getDoc().getCursor().line;
         if (lineNumber !== 0) {
           return false;
         }
 
-        this.setState((state) => {
-          const newCursor = Math.min(
-            state.commandCursor + 1,
-            state.commandHistory.length - 1
-          );
-          this._cm.getDoc().setValue(state.commandHistory[newCursor] || '');
-          const cursorPos = this._cm.getDoc().getLine(0).length - 1;
-          this._cm.getDoc().setCursor({ line: 0, ch: cursorPos });
-          return { commandCursor: newCursor };
+        setCommandCursor((prevCursor) => {
+          const newCursor = Math.min(prevCursor + 1, commandHistory.length - 1);
+          cm.getDoc().setValue(commandHistory[newCursor] || '');
+          const cursorPos = cm.getDoc().getLine(0).length - 1;
+          cm.getDoc().setCursor({ line: 0, ch: cursorPos });
+          return newCursor;
         });
       } else if (e.key === 'ArrowDown') {
-        const lineNumber = this._cm.getDoc().getCursor().line;
-        const lineCount = this._cm.getValue().split('\n').length;
+        const lineNumber = cm.getDoc().getCursor().line;
+        const lineCount = cm.getValue().split('\n').length;
         if (lineNumber + 1 !== lineCount) {
           return false;
         }
 
-        this.setState((state) => {
-          const newCursor = Math.max(state.commandCursor - 1, -1);
-          this._cm.getDoc().setValue(state.commandHistory[newCursor] || '');
-          const newLineCount = this._cm.getValue().split('\n').length;
-          const newLine = this._cm.getDoc().getLine(newLineCount);
+        setCommandCursor((prevCursor) => {
+          const newCursor = Math.max(prevCursor - 1, -1);
+          cm.getDoc().setValue(commandHistory[newCursor] || '');
+          const newLineCount = cm.getValue().split('\n').length;
+          const newLine = cm.getDoc().getLine(newLineCount);
           const cursorPos = newLine ? newLine.length - 1 : 1;
-          this._cm.getDoc().setCursor({ line: lineCount, ch: cursorPos });
-          return { commandCursor: newCursor };
+          cm.getDoc().setCursor({ line: lineCount, ch: cursorPos });
+          return newCursor;
         });
       }
       return true;
-    });
+    };
 
-    this._cm.getWrapperElement().style[
+    codemirrorRef.current.on('keydown', handleKeyDown);
+    codemirrorRef.current.getWrapperElement().style[
       'font-size'
-    ] = `${this.props.fontSize}px`;
-  }
+    ] = `${fontSize}px`;
 
-  componentDidUpdate(prevProps) {
-    this._cm.setOption('theme', `p5-${this.props.theme}`);
-    this._cm.getWrapperElement().style[
-      'font-size'
-    ] = `${this.props.fontSize}px`;
-    this._cm.refresh();
-  }
+    return () => {
+      codemirrorRef.current.off('keydown', handleKeyDown);
+      codemirrorRef.current = null;
+    };
+  }, [theme, dispatchConsoleEvent, fontSize]);
```

Comment:
We have to be careful here because we don't want to re-create the entire codemirror instance when `theme` or `fontSize` change.

I think (need to try it out to be sure) that you can remove the `theme: `p5-${theme}`,` and the setting of the fontSize from this main `useEffect` as they are already covered by your next `useEffect`.